ext/ddtrace_profiling_native_extension/stack_recorder.c

```diff
@@ -436,3 +438,17 @@ static ddog_Timespec time_now() {
 
   return (ddog_Timespec) {.seconds = current_time.tv_sec, .nanoseconds = (uint32_t) current_time.tv_nsec};
 }
+
+static VALUE _native_clear(DDTRACE_UNUSED VALUE _self, VALUE recorder_instance) {
+  struct stack_recorder_state *state;
+  TypedData_Get_Struct(recorder_instance, struct stack_recorder_state, &stack_recorder_typed_data, state);
+
+  ddog_Timespec finish_timestamp = time_now();
+
+  ddog_Profile *profile = serializer_flip_active_and_inactive_slots(state, finish_timestamp);
```

Comment:
Why does the slot need to be flipped here?

Reply:
That is a good question. The answer is... they don't strictly need to be right now, but doing so leaves us with some flexibility for the future.

The more specific answer is: the `StackRecorder` is built in such a way that it can serialize (and now clear) a profile even while another thread from the profiler is writing profiling data to the profile.

Thus, "theoretically" the slot flipping is needed to retain this property -- the other slot by definition is kept empty so the other thread can start writing to it even while we're cleaning the previously-active profile (which is the one that will have leftover data after the fork).

BUT in practice right now writing data to the profile requires the global VM lock, and `_native_clear` is only executed with the global VM lock acquired, so right now there can be no concurrency between writing and clearing. So we could just clear the currently-active profile and keep that as the active, rather than doing the flipping. But that would create potential problems when (if?), in the future, we move more of the "writing to profile" process to be executed outside of the GVL.

Reply:
Added a note about this in 96acfb253f84fc78778f6ea62a26a5a591a6dc17

Reply:
Makes sense, I think the explanation you added in https://github.com/DataDog/dd-trace-rb/commit/96acfb253f84fc78778f6ea62a26a5a591a6dc17 clarifies it a lot! 🙇